controllers/user.js

```diff
@@ -1,32 +1,58 @@
 const { User, Token } = require("../models");
 const { randomBytes } = require("crypto");
+const bcrypt = require("bcrypt");
+const BackendError = require("../utils/BackendError");
 
 const createToken = (user) => {
-  return Token({
+  return new Token({
     user: user._id,
     token: randomBytes(40).toString("hex"),
   });
 };
 
-const login = async (req, res) => {
-  // TODO: Read username, pwd from the req object
-  // Check if data is valid
-  // Return correct status codes: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes
-  // If the user is verified, then return a token along with correct status code
+const login = async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    if (username == null) throw new BackendError(400, "Username is required");
+    if (!username) throw new BackendError(400, "Username cannot be blank");
+    if (password == null) throw new BackendError(400, "Password is required");
+    if (!password) throw new BackendError(400, "Password cannot be blank");
+    const user = await User.findOne({ username });
+    const result = await bcrypt.compare(password, user.password).catch(next);
+    if (!result) {
+      throw new BackendError(400, "Either username or password is incorrect");
+    }
+    const token = await Token.findOne({ user: user._id });
+    return res.status(200).json({ token: token.token });
+  } catch (err) {
+    next(err);
+  }
 };
 
-const signup = async (req, res) => {
-  // TODO: Read username, email, name, pwd from the req object
-  // Hash the password
-  // Return with appropriate status code in case of an error
-  // If successful, return with an appropriate token along with correct status code
+const signup = async (req, res, next) => {
+  try {
+    const { username, email, name, password } = req.body;
+    if (username == null) throw new BackendError(400, "Username is required");
+    if (!username) throw new BackendError(400, "Username cannot be blank");
```

Comment:
combine them into one. Also, "   " will pass both the tests. Be sure to check these cases.